### PR TITLE
feat(hyper): switch to std::io, std::net, and std::path.

### LINKED ISSUE #347: Switch to new std::io
Installed Rust nightly build, after not using Rust for 3 weeks. Hyper won't compile any more.
The I/O libraries changed. Again. Not your fault. Sigh.

```
Compiling hyper v0.2.1
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:215:25: 215:29 error: mismatched types:
 expected `&std::path::Path`,
    found `&std::old_path::posix::Path`
(expected struct `std::path::Path`,
    found struct `std::old_path::posix::Path`) [E0308]
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:215                         cert, X509FileType::PEM).map(lift_ssl_error));
                                                                                                              ^~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:19:1: 24:2 note: in expansion of try_some!
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:214:17: 215:71 note: expansion site
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:217:25: 217:28 error: mismatched types:
 expected `&std::path::Path`,
    found `&std::old_path::posix::Path`
(expected struct `std::path::Path`,
    found struct `std::old_path::posix::Path`) [E0308]
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:217                         key, X509FileType::PEM).map(lift_ssl_error));
                                                                                                              ^~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:19:1: 24:2 note: in expansion of try_some!
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:216:17: 217:70 note: expansion site
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:243:23: 243:57 error: the trait `std::io::Read` is not implemented for the type `std::old_io::net::tcp::TcpStream` [E0277]
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:243                 match SslStream::<TcpStream>::new_server(&**ssl_context, stream) {
                                                                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:243:23: 243:57 error: the trait `std::io::Write` is not implemented for the type `std::old_io::net::tcp::TcpStream` [E0277]
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:243                 match SslStream::<TcpStream>::new_server(&**ssl_context, stream) {
                                                                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:249:42: 249:48 error: attempted access of field `desc` on type `&std::io::error::Error`, but no field with that name was found
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:249                             detail: Some(e.desc.to_string())
                                                                                                                               ^~~~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:289:55: 289:64 error: type `&mut openssl::ssl::SslStream<std::old_io::net::tcp::TcpStream>` does not implement any method in scope named `read`
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:289             HttpStream::Https(ref mut inner) => inner.read(buf)
                                                                                                                                            ^~~~~~~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:299:55: 299:69 error: type `&mut openssl::ssl::SslStream<std::old_io::net::tcp::TcpStream>` does not implement any method in scope named `write_all`
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:299             HttpStream::Https(ref mut inner) => inner.write_all(msg)
                                                                                                                                            ^~~~~~~~~~~~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:306:55: 306:62 error: type `&mut openssl::ssl::SslStream<std::old_io::net::tcp::TcpStream>` does not implement any method in scope named `flush`
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:306             HttpStream::Https(ref mut inner) => inner.flush(),
                                                                                                                                            ^~~~~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:315:55: 315:64 error: type `&mut openssl::ssl::SslStream<std::old_io::net::tcp::TcpStream>` does not implement any method in scope named `get_mut`
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:315             HttpStream::Https(ref mut inner) => inner.get_mut().peer_name()
                                                                                                                                            ^~~~~~~~~
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:346:35: 346:49 error: the trait `std::io::Read` is not implemented for the type `std::old_io::net::tcp::TcpStream` [E0277]
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:346                 let stream = try!(SslStream::new(&context, stream).map_err(lift_ssl_error));
                                                                                                                        ^~~~~~~~~~~~~~
<std macros>:1:1: 6:57 note: in expansion of try!
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:346:30: 346:93 note: expansion site
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:346:35: 346:49 error: the trait `std::io::Write` is not implemented for the type `std::old_io::net::tcp::TcpStream` [E0277]
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:346                 let stream = try!(SslStream::new(&context, stream).map_err(lift_ssl_error));
                                                                                                                        ^~~~~~~~~~~~~~
<std macros>:1:1: 6:57 note: in expansion of try!
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:346:30: 346:93 note: expansion site
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:363:29: 363:32 error: mismatched types:
 expected `std::old_io::IoError`,
    found `std::io::error::Error`
(expected struct `std::old_io::IoError`,
    found struct `std::io::error::Error`) [E0308]
/home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.2.1/src/net.rs:363         StreamError(err) => err,
                                                                                                                  ^~~
error: aborting due to 12 previous errors
Could not compile `hyper`.
```

### PR DESCRIPTION
**Pull request by seanmonstar**
All instances of `old_io` and `old_path` were switched to use the new
shiny `std::io`, `std::net`, and `std::path` modules. This means that
`Request` and `Response` implement `Read` and `Write` now.

Because of the changes to `TcpListener`, this also takes the opportunity
to correct the method usage of `Server`. As with other
languages/frameworks, the server is first created with a handler, and
then a host/port is passed to a `listen` method. This reverses what
`Server` used to do.

Closes #347

BREAKING CHANGE: Check the docs. Everything was touched.

### CODE REVIEW

**Comment by reem** [comment]
Needs a rebase already.

File: Cargo.toml [hunk 1]
```diff
@@ -21,3 +21,6 @@ time = "*"
 unicase = "*"
 unsafe-any = "*"
 url = "*"
+
+[profile.bench]
+opt-level = 0
```
Comment (reem) [inline Cargo.toml]: ??

Comment (seanmonstar) [inline Cargo.toml]: Oh dangit. I ahead set that when I was fixing the bench marks, to compile faster.

**Comment by seanmonstar** [comment]
@reem rebased.

Interesting to note: speed improvements with the `client_mock_tcp` bench is over 500%.

```
BEFORE
test bench_mock_hyper ... bench:    280059 ns/iter (+/- 62609)

AFTER
test bench_mock_hyper ... bench:     50152 ns/iter (+/- 5771)
```

File: src/http.rs [hunk 2]
```diff
@@ -65,13 +65,13 @@ impl<R: Reader> HttpReader<R> {
...
-impl<R: Reader> Reader for HttpReader<R> {
-    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
+impl<R: Read> Read for HttpReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         match *self {
             SizedReader(ref mut body, ref mut remaining) => {
                 debug!("Sized read, remaining={:?}", remaining);
                 if *remaining == 0 {
-                    Err(old_io::standard_error(old_io::EndOfFile))
+                    Ok(0)
```
Comment (reem) [inline src/http.rs]: I think `Ok(0)` is still bad behavior, unless the recent RFC changed that.

File: src/http.rs [hunk 3]
```diff
@@ -266,29 +287,24 @@ impl<W: Writer> Writer for HttpWriter<W> {
...
-                } else {
-                    Err(old_io::IoError {
-                        kind: old_io::ShortWrite(bytes),
-                        desc: "EmptyWriter cannot write any bytes",
-                        detail: Some("Cannot include a body with this kind of message".to_string())
-                    })
+                if msg.len() != 0 {
+                    error!("Cannot include a body with this kind of message");
                 }
+                Ok(0)
```
Comment (reem) [inline src/http.rs]: Same `Ok(0)` note here.

File: src/http.rs [hunk 4]
```diff
@@ -345,24 +361,36 @@ pub fn is_token(b: u8) -> bool {
 /// otherwise returns any error encountered reading the stream.
 ///
 /// The remaining contents of `buf` are left untouched.
-fn read_token_until_space<R: Reader>(stream: &mut R, buf: &mut [u8]) -> HttpResult<bool> {
-    use std::old_io::BufWriter;
-    let mut bufwrt = BufWriter::new(buf);
+fn read_token_until_space<R: Read>(stream: &mut R, buf: &mut [u8]) -> HttpResult<bool> {
```
Comment (reem) [inline src/http.rs]: can this be renamed, since it is specialized to methods?

File: src/http.rs [hunk 5]
```diff
@@ -507,14 +549,24 @@ pub type RawHeaderLine = (String, Vec<u8>);
 /// >                ; obsolete line folding
 /// >                ; see Section 3.2.4
 /// > ```
-pub fn read_header<R: Reader>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
+pub fn read_header<R: Read>(stream: &mut R) -> HttpResult<Option<RawHeaderLine>> {
+    macro_rules! byte (
```
Comment (reem) [inline src/http.rs]: This whole macro business is honestly horrible, why is this necessary? Can't we add an extension method?

File: src/http.rs (hunk 2 above)
Comment (seanmonstar) [inline src/http.rs]: My understanding is that Ok(0) is EOF. There no longer is an EOF error kind...

File: src/http.rs (hunk 5 above)
Comment (seanmonstar) [inline src/http.rs]: Yea, it's gross. The change from an EOF error kind to Ok(0) makes this harder. I tried a ReadByte extension trait, but could propagate the correct error upon EOF. So, the macros allow inserting the appropriate HttpError when EOF is found early. 

I also have been working on a branch to use httparse, which removes much of this module.

Comment (reem) [inline src/http.rs]: Just found httpparse, seems interesting but we'll need a push parser for async integration later.

Comment (seanmonstar) [inline src/http.rs]: httparse is a push parser. My goal is to be like Pico, but using Iterator,
to keep things safe.

File: src/http.rs (hunk 2 above)
Comment (reem) [inline src/http.rs]: Asked about this and you're right.

File: src/http.rs (hunk 5 above)
Comment (reem) [inline src/http.rs]: Oh, so the idea is I call `parse` many times?

Comment (seanmonstar) [inline src/http.rs]: Yes, but just like Pico, it's stateless. It will update the `Request` object with all the info it could find, but it won't keep track of it's state (as Pico's developer points out, keeping state makes it slower). So you can read some bytes, call `parse()`, and if it came back as `Status::Impartial` (or `-2` in Pico), you can still inspect some properties. Perhaps you check the route and can determine it's a 404 early before having to parse any more. You should, however, read more bytes into the same buffer, and call `parse()` again. The parser will start from the beginning.

It's faster this way because: keeping state (and checking it every single byte) is slower, and in **most** cases, reading from a socket should read the entire head, so you shouldn't often get `Partial` parses.

**Comment by seanmonstar** [comment]
If there's no real problems, I'd like to merge so we can unbreak people on crates.io...

**Comment by akoskaaa** [comment]
Hey, I am not sure if this is a blocker, but with the current build the example client does not build for me. Here's the exception:
```
src/main.rs:19:20: 19:36 error: type `hyper::client::response::Response` does not implement any method in scope named `read_to_string`
src/main.rs:19     let body = res.read_to_string().unwrap();
                                  ^~~~~~~~~~~~~~~~
src/main.rs:19:36: 19:36 help: methods from traits can only be called if the trait is in scope; the following trait is implemented but not in scope, perhaps add a `use` for it:
src/main.rs:19:36: 19:36 help: candidate #1: use `std::io::Read`
error: aborting due to previous error
```

**Comment by maximih** [comment]
@akoskaaa read_to_string() was changed in the new io mdoule. Looks like the example needs to be updated. Here is how I modified my code to make it compile:
        
        use std::io::Read;
        ...
        let mut content: String = "".to_string();
        try!(response.read_to_string(&mut content));

**Comment by seanmonstar** [comment]
Alright, I updated the REAME to work, and renamed that odd function. Once travis hits green, I'll merge. (sorry those using git rev's, I probably broke you because of my force-push)
